cloudigrade/analyzer/management/commands/analyze_log.py

```diff
@@ -93,26 +98,55 @@ def _parse_log_for_ec2_events(self, log):
                 .get('instancesSet', {})\
                 .get('items', [])
 
+            # Collect the EC2 instances the API was called on
             for item in ec2_info:
                 instance_id = item.get('instanceId')
-                instance_type = item.get('instanceType')
-                ec2_ami_id = item.get('imageId')
-                subnet = item.get('subnetId')
-
-                instances.append((account_id, instance_id, region))
-
-                # Subnet, ec2_ami_id, and instance_type will be None
-                # for start/stop events.
-                instance_events.append(
-                    (instance_id,
-                     event_type,
-                     occured_at,
-                     subnet,
-                     ec2_ami_id,
-                     instance_type)
-                )
+                instances[instance_id]['account_id'] = account_id
+                instances[instance_id]['region'] = region
+
+                event = self._build_instance_event(account_id, region, item)
+                # Event type and time are the same for each instance
+                event['event_type'] = event_type
+                event['occurred_at'] = occured_at
+
+                if instances[instance_id].get('events'):
+                    instances[instance_id]['events'].append(event)
+                else:
+                    instances[instance_id]['events'] = [event]
+
+        return dict(instances)
+
+    def _build_instance_event(self, account_id, region, log_record):
+        """
+        Return critical data for an instance on/off event.
+
+        Args:
+            account_id (str): The account_id the instance belongs to.
+            region (str): The AWS region the instance resides in.
+            log_record (str): The instance specific data from the log record.
 
-        return list(set(instances)), instance_events
+        Returns:
+            dict: EC2 instance data for the event.
+
+        """
+        ec2_ami_id = log_record.get('imageId')
+        instance_id = log_record.get('instanceId')
+        instance_type = log_record.get('instanceType')
+        subnet = log_record.get('subnetId')
+
+        if subnet is None \
```

Comment:
If any of them are `None`, we squash _all_ of them?

Is there ever a case where we'd get and want to keep some from the `log_record`? Or is it safe to assume that if any are missing, they're probably all missing and we need fetch from the instance?

Reply:
Realistically, I think if any of them is `None` all of them are. Those values _can't_ be none on an active EC2 instance, so we'll need to get the data regardless.